pkg/validation/validation.go

```diff
@@ -0,0 +1,138 @@
+	}
+	currentVersion, err := semver.Parse(cv)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse current version %s as semver: %v", cv, err)
+	}
+
+	// Compare versions to ascertain if upgrade should proceed.
+	proceed := compareVersions(desiredVersion, currentVersion, logger)
+	if !proceed {
+		return false, nil
+	}
+	logger.Info(fmt.Sprintf("Desired version %s validated as greater then current version %s", desiredVersion, currentVersion))
+
+	// Validate available version is in Cincinnati.
+	desiredChannel := uC.Spec.Desired.Channel
+	clusterId, err := uuid.Parse(string(cV.Spec.ClusterID))
+	if err != nil {
+		return false, err
+	}
+	upstreamURI, err := url.Parse(string(cV.Spec.Upstream))
+	if err != nil {
+		return false, err
+	}
+
+	updates, err := cincinnati.NewClient(clusterId, nil, nil).GetUpdates(upstreamURI, runtime.GOARCH, desiredChannel, currentVersion)
+	if err != nil {
+		return false, err
+	}
+
+	var cvoUpdates []configv1.Update
+	for _, update := range updates {
+		cvoUpdates = append(cvoUpdates, configv1.Update{
+			Version: update.Version.String(),
+			Image:   update.Image,
+		})
+	}
+
+	// Check whether the desired version exists in availableUpdates
+	found := false
+	for _, v := range cvoUpdates {
+		if v.Version == dv && !v.Force {
+			found = true
+		}
+	}
+
+	if !found {
+		logger.Info(fmt.Sprintf("Failed to find the desired version %s in channel %s", desiredVersion, desiredChannel))
+		//We need to update the condition
+		errMsg := fmt.Sprintf("cannot find version %s in available updates", desiredVersion)
+		return false, fmt.Errorf(errMsg)
+	}
+	return true, nil
+}
+
+// compareVersions accepts desiredVersion and currentVersion strings as versions, converts
+// them to semver and then compares them. Returning true only if desiredVersion > currentVersion.
+func compareVersions(dV semver.Version, cV semver.Version, logger logr.Logger) bool {
+	result := dV.Compare(cV)
+	switch result {
+	case -1:
+		logger.Info(fmt.Sprintf("%s is less then %s", dV, cV))
+		logger.Info(fmt.Sprintf("Downgrading cluster is not supported. Not Proceeding to %s", dV))
+		return false
+	case 0:
+		logger.Info(fmt.Sprintf("%s is equal to %s", dV, cV))
+		logger.Info(fmt.Sprintf("Cluster is already on version %s", cV))
+		return false
+	case 1:
+		logger.Info(fmt.Sprintf("%s is greater then %s", dV, cV))
+	}
+	return true
+
+}
+
+//go:generate mockgen -destination=mockValidationBuilder.go -package=validation github.com/openshift/managed-upgrade-operator/pkg/validation ValidationBuilder
+type ValidationBuilder interface {
+	NewClient(client client.Client) (Validator, error)
+}
+
+// validationBuilder is an empty struct that enables instantiation of this type and its
+// implimented interface.
+type validationBuilder struct{}
+
+// NewClient returns a Validator interface or an error if one occurs.
+func (vb *validationBuilder) NewClient(client client.Client) (Validator, error) {
```

Comment:
The constructor doesn't seem to be saving the kube client to the struct or using it, is it needed?
Could either remove the client dependency (unless you foresee a need for it in future revisions), or alternately keep it and store it and also maybe move the GetClusterVersion() call (which uses the client) out of the controller and into the IsValidUpgradeConfig() function.

Reply:
Resolved dep: ad2e4705621ded503b2b5a67cd942055fff39b2c